src/mesh/Data.cpp

```diff
@@ -78,5 +82,52 @@ void Data::resetDataCount()
   _dataCount = 0;
 }
 
+void Data::storeDataInWaveform(int waveformSampleID)
+{
+  PRECICE_TRACE();
+  PRECICE_ASSERT(_ptrWaveform->valuesSize() == _values.size(),
+                 _ptrWaveform->valuesSize(), _values.size());
+  _ptrWaveform->storeAt(_values, waveformSampleID);
+}
+
+int Data::sizeOfSampleStorageInWaveform()
+{
+  PRECICE_TRACE();
+  return _ptrWaveform->sizeOfSampleStorage();
+}
+
+Eigen::VectorXd Data::waveformSampleAt(double normalizedDt)
+{
+  PRECICE_TRACE();
+  PRECICE_ASSERT(_ptrWaveform->valuesSize() == _values.size(), _ptrWaveform->valuesSize(), _values.size());
+  PRECICE_ASSERT(normalizedDt >= 0, "Sampling outside of valid range!");
+  PRECICE_ASSERT(normalizedDt <= 1, "Sampling outside of valid range!");
+  return _ptrWaveform->sample(normalizedDt);
+}
+
+void Data::initializeWaveform()
+{
+  PRECICE_TRACE();
+  // PRECICE_ASSERT(_values.size() > 0, _values.size());  // @todo assertion breaks, but seems like calling advance on empty write data is ok?
```

Comment:
I moved this assertion in 9a71ca0ca196bfdd59f4c96e780f28c8df4e151d. And yes: Some tests are breaking. I have to investigate this furtherly. Might also be a bug in the implementation of the waveform relaxation.

Reply:
To provide some context here: We don't call `advance` before `initialize`, but the mesh is empty (no `setMeshXXX(...)` called on one participant). I think in some cases this is valid, in other cases it is not. This assertion seems too strict from my point of view, because an empty mesh is allowed for parallel runs, where individual ranks do not define a mesh.

I opened the issue #1158, fixed the serial tests and removed this assertion in 67a3f751577531610e215b32c791938763a32319. I think this resolves this discussion.